### PR TITLE
[wip] subscription cleanup

### PR DESCRIPTION
**Pull request by divagant-martian**
## Description

WIP

## Breaking Changes

n/a

## Notes & open questions

n/a

## Change checklist

- [ ] Self-review.
- [ ] Documentation updates following the [style guide](https://rust-lang.github.io/rfcs/1574-more-api-documentation-conventions.html#appendix-a-full-conventions-text), if relevant.
- [ ] Tests if relevant.
- [ ] All breaking changes documented.

### CODE REVIEW

File: src/net.rs [hunk 1]
```diff
@@ -636,6 +770,26 @@ impl Actor {
...
+        if let Entry::Occupied(mut entry) = self.topics.entry(topic) {
+            let state = entry.get_mut();
+            state.event_senders.remove(&receiver_id);
+            if !state.still_needed() {
+                self.quit_queue.push_back(topic);
+                self.process_quit_queue().await?;
+            }
+        } else {
+            // topic should not have been dropped without all receivers being dropped first
+            warn!("receiver gone for missing topic");
```
Comment (dignifiedquire) [inline src/net.rs:788]: maybe log the topic, so there is a chance to figure out what's going on

**Comment by dignifiedquire** [comment]
I believe this doesn't fix the issue in `connection_loop` where read or write errors kill the whole connection loop, because of the way `try_join` works

File: src/net.rs [hunk 2]
```diff
@@ -236,6 +226,64 @@ impl Gossip {
...
+}
+
+impl Drop for EventStream {
+    fn drop(&mut self) {
+        // NOTE: unexpectedly, this works without a tokio runtime, so we leverage that to avoid yet
+        // another spawned task
+        if let Err(e) = self.to_actor_tx.try_send(ToActor::ReceiverGone {
+            topic: self.topic,
+            receiver_id: self.receiver_id,
+        }) {
```
Comment (matheus23) [inline src/net.rs:266]: You could actually use [`to_actor_tx.reserve_owned`](https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Sender.html#method.reserve_owned) and store the `OwnedPermit` in `EventStream`.

This way you'll guarantee that you can always have space for this message and sending it will always be a synchronous call.

**Comment by matheus23** [comment]
(sorry for the so far borderline useless review comments, will continue reviewing later)

File: src/net.rs (hunk 2 above)
Comment (divagant-martian) [reply src/net.rs:266]: I'm not sure this is a good idea. The actor channel is (correctly) bounded, with just a capacity of 64. So this would top our maximum subscriptions to that amount and block any other message to the actor since the permit is long lived (from subscription start to end)

File: src/net.rs [hunk 3]
```diff
@@ -236,6 +226,64 @@ impl Gossip {
...
+#[derive(derive_more::Debug)]
+pub struct EventStream {
+    /// The actual stream polled to return [`Event`]s to the application.
+    #[debug("Stream")]
+    inner: Pin<Box<dyn Stream<Item = Result<Event>> + Send + 'static>>,
+
+    /// Channel to the actor task.
+    ///
+    /// This is used to handle the receiver being dropped. When all receiver and publishers are
+    /// gone the topic will be desubscribed.
```
[the remark this answers is not in the document]
Comment (divagant-martian) [reply src/net.rs]: done in 6138e47

File: src/net.rs (hunk 1 above)
Comment (divagant-martian) [reply src/net.rs]: done in 6138e47

File: src/net.rs (hunk 2 above)
Comment (matheus23) [reply src/net.rs:266]: Good point.

File: src/net.rs [hunk 4]
```diff
@@ -278,98 +330,180 @@ struct Actor {
...
-                        self.handle_in_event(InEvent::TimerExpired(timer), now).await.context("timers.drain_expired -> handle_in_event")?;
+            }
+            event = self.in_event_rx.recv() => {
+                trace!(?i, "tick: in_event_rx");
+                inc!(Metrics, actor_tick_in_event_rx);
+                match event {
+                    Some(event) => {
+                        self.handle_in_event(event, Instant::now()).await.context("in_event_rx.recv -> handle_in_event")?;
                     }
+                    None => unreachable!()
```
Comment (rklaehn) [inline src/net.rs:476]: why is this unreachable? It it impossible for this channel to be closed?

File: src/net.rs (hunk 2 above)
Comment (rklaehn) [reply src/net.rs:266]: Something like a channel that is bounded for normal messages but unbounded for special shutdown messages would be nice sometimes...

**Review by rklaehn — APPROVED** [review]
Seems like a good improvement overall...

File: src/net.rs [hunk 5]
```diff
@@ -278,98 +330,180 @@ struct Actor {
...
                     }
-                }
-                event = self.in_event_rx.recv() => {
-                    trace!(?i, "tick: in_event_rx");
-                    inc!(Metrics, actor_tick_in_event_rx);
-                    match event {
-                        Some(event) => {
-                            self.handle_in_event(event, Instant::now()).await.context("in_event_rx.recv -> handle_in_event")?;
-                        }
-                        None => unreachable!()
```
Comment (divagant-martian) [inline src/net.rs:352]: @rklaehn this is the original line, unchanged. Since the code moved in a way such that it's now with a a different indentation it's a bit hard to parse. To specifically answer your question I don't really know

Comment (rklaehn) [reply src/net.rs:352]: yeah, I was just curious. I saw that this wasn't your code that introduced this...
